File: src/main/java/de/tsenger/vdstools/vds/DigitalSeal.java
package de.tsenger.vdstools.vds;

import java.io.IOException;
import java.math.BigInteger;
import java.nio.ByteBuffer;
import java.time.LocalDate;
import java.util.List;

import org.bouncycastle.util.Arrays;
import org.bouncycastle.util.encoders.Hex;
import org.tinylog.Logger;

import de.tsenger.vdstools.DataEncoder;
import de.tsenger.vdstools.DataParser;
import de.tsenger.vdstools.DerTlv;

/**
 * @author Tobias Senger
 *
 */
public class DigitalSeal {

	private String vdsType;
	private VdsHeader vdsHeader;
	private VdsMessage vdsMessage;
	private VdsSignature vdsSignature;

	public DigitalSeal(VdsHeader vdsHeader, VdsMessage vdsMessage, VdsSignature vdsSignature) {
		this.vdsHeader = vdsHeader;
		this.vdsMessage = vdsMessage;
		this.vdsSignature = vdsSignature;
		this.vdsType = vdsHeader.getVdsType();
	}

	public String getVdsType() {
		return vdsType;
	}

	public String getIssuingCountry() {
		return vdsHeader.issuingCountry;
	}

	/**
	 * Returns a string that identifies the signer certificate. The SignerCertRef
	 * string is build from Signer Identifier (country code || signer id) and
	 * Certificate Reference. The Signer Identifier maps to the signer certificates
	 * subject (C || CN) The Certificate Reference will be interpreted as an hex
	 * string integer that represents the serial number of the signer certificate.
	 * Leading zeros in Certificate Reference the will be cut off. e.g. Signer
	 * Identifier 'DETS' and CertificateReference '00027' will result in 'DETS27'
	 * 
	 * @return Formated SignerCertRef all UPPERCASE
	 */
	public String getSignerCertRef() {
		BigInteger certRefInteger = new BigInteger(vdsHeader.certificateReference, 16);
		return String.format("%s%x", vdsHeader.signerIdentifier, certRefInteger).toUpperCase();
	}

	public String getSignerIdentifier() {
		return vdsHeader.signerIdentifier;
	}

	public String getCertificateReference() {
		return vdsHeader.certificateReference;
	}

	public LocalDate getIssuingDate() {
		return vdsHeader.issuingDate;
	}

	public LocalDate getSigDate() {
		return vdsHeader.sigDate;
	}

	public byte getDocFeatureRef() {
		return vdsHeader.docFeatureRef;
	}

	public byte getDocTypeCat() {
		return vdsHeader.docTypeCat;
	}

	public byte[] getHeaderAndMessageBytes() {
		return Arrays.concatenate(vdsHeader.getEncoded(), vdsMessage.getEncoded());
	}

	public byte[] getEncoded() throws IOException {
		return Arrays.concatenate(vdsHeader.getEncoded(), vdsMessage.getEncoded(), vdsSignature.getEncoded());
	}

	public byte[] getSignatureBytes() {
		return vdsSignature.getPlainSignatureBytes();
	}

	public String getRawString() throws IOException {
		return DataEncoder.encodeBase256(getEncoded());
	}

	public <T> T getFeature(String feature) {
		return vdsMessage.getDocumentFeature(feature);
	}

	public <T> void addFeature(String feature, T value) {
		vdsMessage.addDocumentFeature(feature, value);
	}

	public static DigitalSeal fromRawString(String rawString) {
		DigitalSeal seal = null;
		try {
			seal = parseVdsSeal(DataParser.decodeBase256(rawString));
		} catch (IOException e) {
			Logger.error(e.getMessage());
		}
		return seal;
	}

	public static DigitalSeal fromByteArray(byte[] rawBytes) {
		DigitalSeal seal = null;
		try {
			seal = parseVdsSeal(rawBytes);
		} catch (IOException e) {
			Logger.error(e.getMessage());
		}
		return seal;
	}

	private static DigitalSeal parseVdsSeal(byte[] rawBytes) throws IOException {

		ByteBuffer rawData = ByteBuffer.wrap(rawBytes);
		Logger.trace("rawData: {}", () -> Hex.toHexString(rawBytes));

		VdsHeader vdsHeader = VdsHeader.fromByteBuffer(rawData);
		VdsMessage vdsMessage = new VdsMessage(vdsHeader.getVdsType());
		VdsSignature vdsSignature = null;

		int messageStartPosition = rawData.position();

		List<DerTlv> derTlvList = DataParser
				.parseDerTLvs(Arrays.copyOfRange(rawBytes, messageStartPosition, rawBytes.length));

		for (DerTlv derTlv : derTlvList) {
			if (derTlv.getTag() == (byte) 0xff) {
				vdsSignature = VdsSignature.fromByteArray(derTlv.getEncoded());
			} else {
				vdsMessage.addDerTlv(derTlv);
			}
		}
		return new DigitalSeal(vdsHeader, vdsMessage, vdsSignature);

	}

}

File: src/main/java/de/tsenger/vdstools/vds/DigitalSeal.kt
package de.tsenger.vdstools.vds


import co.touchlab.kermit.Logger
import de.tsenger.vdstools.DataEncoder
import de.tsenger.vdstools.DataParser
import de.tsenger.vdstools.Signer
import de.tsenger.vdstools.asn1.DerTlv
import kotlinx.datetime.LocalDate
import okio.Buffer
import java.io.IOException


class DigitalSeal {
    val vdsType: String
    private val vdsHeader: VdsHeader
    private val vdsMessage: VdsMessage
    private val vdsSignature: VdsSignature?

    private constructor(vdsHeader: VdsHeader, vdsMessage: VdsMessage, vdsSignature: VdsSignature?) {
        this.vdsHeader = vdsHeader
        this.vdsMessage = vdsMessage
        this.vdsSignature = vdsSignature
        this.vdsType = vdsHeader.vdsType
    }

    constructor(vdsHeader: VdsHeader, vdsMessage: VdsMessage, signer: Signer) {
        this.vdsHeader = vdsHeader
        this.vdsMessage = vdsMessage
        this.vdsSignature = createVdsSignature(vdsHeader, vdsMessage, signer)
        this.vdsType = vdsHeader.vdsType
    }

    val issuingCountry: String?
        get() = vdsHeader.issuingCountry

    val signerCertRef: String
        get() = vdsHeader.signerCertRef

    val signerIdentifier: String?
        get() = vdsHeader.signerIdentifier

    val certificateReference: String?
        get() = vdsHeader.certificateReference

    val issuingDate: LocalDate?
        get() = vdsHeader.issuingDate

    val sigDate: LocalDate?
        get() = vdsHeader.sigDate

    val docFeatureRef: Byte
        get() = vdsHeader.docFeatureRef

    val docTypeCat: Byte
        get() = vdsHeader.docTypeCat

    val headerAndMessageBytes: ByteArray
        get() = vdsHeader.encoded + vdsMessage.encoded

    @get:Throws(IOException::class)
    val encoded: ByteArray
        get() = vdsHeader.encoded + vdsMessage.encoded + (vdsSignature?.encoded ?: ByteArray(0))

    val signatureBytes: ByteArray
        get() = vdsSignature!!.plainSignatureBytes

    @get:Throws(IOException::class)
    val rawString: String
        get() = DataEncoder.encodeBase256(encoded)

    val featureList: List<Feature>
        get() = vdsMessage.featureList

    fun getFeature(feature: String): Feature? {
        return vdsMessage.getFeature(feature)
    }

    private fun createVdsSignature(vdsHeader: VdsHeader, vdsMessage: VdsMessage, signer: Signer): VdsSignature? {
        val headerMessage = vdsHeader.encoded + vdsMessage.encoded
        try {
            val signatureBytes = signer.sign(headerMessage)
            return VdsSignature(signatureBytes)
        } catch (e: Exception) {
            Logger.e("Signature creation failed: " + e.localizedMessage)
            return null
        }
    }


    companion object {

        @JvmStatic
        fun fromRawString(rawString: String): DigitalSeal? {
            var seal: DigitalSeal? = null
            try {
                seal = parseVdsSeal(DataParser.decodeBase256(rawString))
            } catch (e: Exception) {
                Logger.e(e.localizedMessage)
            }
            return seal
        }

        @JvmStatic
        fun fromByteArray(rawBytes: ByteArray): DigitalSeal? {
            var seal: DigitalSeal? = null
            try {
                seal = parseVdsSeal(rawBytes)
            } catch (e: Exception) {
                Logger.e(e.localizedMessage)
            }
            return seal
        }

        @OptIn(ExperimentalStdlibApi::class)
        @Throws(IOException::class)
        private fun parseVdsSeal(rawBytes: ByteArray): DigitalSeal {
            val rawDataBuffer = Buffer().write(rawBytes)
            Logger.v("rawData: ${rawBytes.toHexString()}")

            val vdsHeader = VdsHeader.fromBuffer(rawDataBuffer)
            var vdsSignature: VdsSignature? = null

            val derTlvList = DataParser
                .parseDerTLvs(rawDataBuffer.readByteArray())

            val featureList: MutableList<DerTlv> = ArrayList(derTlvList.size - 1)

            for (derTlv in derTlvList) {
                if (derTlv.tag == 0xff.toByte()) {
                    vdsSignature = VdsSignature.fromByteArray(derTlv.encoded)
                } else {
                    featureList.add(derTlv)
                }
            }
            val vdsMessage = VdsMessage(vdsHeader.vdsType, featureList)
            return DigitalSeal(vdsHeader, vdsMessage, vdsSignature)
        }
    }
}
